src/irr/asset/CImageLoaderTGA.cpp

```diff
@@ -112,78 +189,133 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
-		
-		if (!flip)
-			out += stride;
+			auto entry = reinterpret_cast<uint8_t*>(inputCreationImage->getBuffer()->getPointer());
+			auto end = entry + inputCreationImage->getBuffer()->getSize();
+			auto stride = trueExtent.X * getTexelOrBlockBytesize(format);
+
+			for (uint32_t y = 0, yRising = 0; y < trueExtent.Y; y += 2, ++yRising)
+				std::swap_ranges(entry + (yRising * stride), entry + ((yRising + 1) * stride), end - ((yRising + 1) * stride));
+		}
 	}
-}
+
+	if (srcFormat == destFormat)
+		newConvertedImage = inputCreationImage;
+	else
+	{
+		using CONVERSION_FILTER = CConvertFormatImageFilter<srcFormat, destFormat>;
+		CONVERSION_FILTER convertFilter;
+		CONVERSION_FILTER::state_type state;
+		{
+			auto referenceImageParams = inputCreationImage->getCreationParameters();
+			auto referenceBuffer = inputCreationImage->getBuffer();
+			auto referenceRegions = inputCreationImage->getRegions();
+			auto referenceRegion = referenceRegions.begin();
+			const auto newTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(destFormat);
+
+			asset::TexelBlockInfo referenceBlockInfo(referenceImageParams.format);
+			core::vector3du32_SIMD referenceTrueExtent = referenceBlockInfo.convertTexelsToBlocks(referenceRegion->getTexelStrides());
+
+			auto newImageParams = referenceImageParams;
+			auto newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(referenceTrueExtent.X * referenceTrueExtent.Y * referenceTrueExtent.Z * newTexelOrBlockByteSize);
+			auto newRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1);
+			newRegions->front() = *referenceRegion;
+
+			newImageParams.format = destFormat;
+			newConvertedImage = ICPUImage::create(std::move(newImageParams));
+			newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
+		}
+
+		auto attachedRegion = newConvertedImage->getRegions().begin();
+
+		state.inImage = inputCreationImage.get();
+		state.outImage = newConvertedImage.get();
+		state.inOffset = { 0, 0, 0 };
+		state.inBaseLayer = 0;
+		state.outOffset = { 0, 0, 0 };
+		state.outBaseLayer = 0;
+		state.extent = attachedRegion->getExtent();
+		state.layerCount = attachedRegion->imageSubresource.layerCount;
+		state.inMipLevel = attachedRegion->imageSubresource.mipLevel;
+		state.outMipLevel = attachedRegion->imageSubresource.mipLevel;
+
+		if (!convertFilter.execute(&state))
+			os::Printer::log("Something went wrong while converting!", ELL_WARNING);
+	}
+
+	return newConvertedImage;
+};
 
 //! creates a surface from the file
 asset::SAssetBundle CImageLoaderTGA::loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
 {
 	STGAHeader header;
-
-	core::smart_refctd_ptr<ICPUBuffer> palette = nullptr;
-
 	_file->read(&header, sizeof(STGAHeader));
 
+	core::smart_refctd_ptr<ICPUImage> colorMap; // not used, but it's texel buffer may be useful in future
```

Comment:
this actually makes far more sense as a raw array of `struct{uint8_t}`, `struct{uint8_t[4]}`, or `struct{uint8_t[4]}` than an actual image

Reply:
basically an `ICPUBuffer` instead of a whole image

Reply:
I did for filters usage

Reply:
fixed in upcoming commit